service/controller/aws/v14/resource/kubeconfig/desired_test.go

```diff
@@ -0,0 +1,177 @@
+kind: Config
+clusters:
+- name: giantswarm-w7utg
+  cluster:
+    server: api.giantswarm.io
+    certificate-authority-data: Y2E=
+users:
+- name: giantswarm-w7utg-user
+  user:
+    client-certificate-data: Y3J0
+    client-key-data: a2V5
+contexts:
+- name: giantswarm-w7utg-context
+  context:
+    cluster: giantswarm-w7utg
+    user: giantswarm-w7utg-user
+current-context: giantswarm-w7utg-context
+preferences: {}
+`
+)
+
+func Test_Resource_GetDesiredState(t *testing.T) {
+	tests := []struct {
+		name            string
+		config          *v1alpha1.AWSClusterConfig
+		expectedSecrets []*corev1.Secret
+		errorMatcher    func(error) bool
+		secretCert      *corev1.Secret
+		timeout         time.Duration
+	}{
+		{
+			name: "case 0: basic match",
+			secretCert: &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"clusterComponent": string(certs.AppOperatorAPICert),
+						"clusterID":        "w7utg",
+					},
+				},
+				Data: map[string][]byte{
+					"ca":  []byte("ca"),
+					"crt": []byte("crt"),
+					"key": []byte("key"),
+				},
+			},
+			config: &v1alpha1.AWSClusterConfig{
+				Spec: v1alpha1.AWSClusterConfigSpec{
+					Guest: v1alpha1.AWSClusterConfigSpecGuest{
+						ClusterGuestConfig: v1alpha1.ClusterGuestConfig{
+							DNSZone: "giantswarm.io",
+							ID:      "w7utg",
+							Name:    "My own snowflake cluster",
+							Owner:   "giantswarm",
+						},
+					},
+				},
+			},
+			expectedSecrets: []*corev1.Secret{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "w7utg-kubeconfig",
+						Namespace: "giantswarm",
+						Labels: map[string]string{
+							"giantswarm.io/cluster":      "w7utg",
+							"giantswarm.io/organization": "giantswarm",
+							"giantswarm.io/managed-by":   "cluster-operator",
+							"giantswarm.io/service-type": "managed",
+						},
+					},
+					Data: map[string][]byte{
+						"kubeConfig": []byte(kubeconfigYaml),
+					},
+				},
+			},
+		},
+		{
+			name: "case 1: cert timeout, reconciliation stop",
+			config: &v1alpha1.AWSClusterConfig{
+				Spec: v1alpha1.AWSClusterConfigSpec{
+					Guest: v1alpha1.AWSClusterConfigSpecGuest{
+						ClusterGuestConfig: v1alpha1.ClusterGuestConfig{
+							DNSZone: "giantswarm.io",
+							ID:      "w7utg",
+							Name:    "My own snowflake cluster",
+							Owner:   "giantswarm",
+						},
+					},
+				},
+			},
+			expectedSecrets: []*corev1.Secret{},
+			timeout:         1 * time.Second,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+
+			client := clientgofake.NewSimpleClientset()
+			fakeWatch := watch.NewFake()
+			client.PrependWatchReactor("secrets", ktesting.DefaultWatchReactor(fakeWatch, nil))
+
+			c := Config{
+				K8sClient: client,
+				Logger:    microloggertest.New(),
+
+				ProjectName:       "cluster-operator",
+				ResourceNamespace: "giantswarm",
+			}
+
+			if tc.timeout != 0 {
+				c.CertsWatchTimeout = tc.timeout
+			}
```

Comment:
I don't really like this setting of the timeout in the test.

It would be better to define a private method that returns the generated secret without using the certs searcher. That code will be much easier to test.

Pawel made this point in the last sig-operator sync and it's a good one.